New "How to route Tasks" guide (#58)
temporalio/documentation: docs/route-tasks.md

---
id: route-tasks
title: How to route Tasks
sidebar_label: Route Tasks
---

Consider the specifications of a Worker:

> - A Worker is meant to handle only the Tasks that it is registered to handle. This means that Workers registered to handle Tasks for Workflow A are not capable of handling Tasks for Workflow B, and vice versa. 

Reviewer: ```suggestion
- A Worker only handles the Tasks that it is registered to handle. This means that Workers registered to handle Tasks for Workflow A are not capable of handling Tasks for Workflow B, and vice versa. 
```

Other than rewording the first sentence, I think the description is a bit confusing. It sort of suggests that a Worker can't handle tasks for both Workflow A and B inherently. Obviously I can have Worker listen on a single taskqueue and support multiple types (A and B).

> - A Worker runs on a speicific host. This means that Activity Tasks that depend on resources specific to a particular host must be executed by a Worker on that host.

Reviewer: ```suggestion
- A Worker runs on a speicific host. This means that Activity Tasks that depend on host-specific resources must be executed by a Worker on that host.
```

I think giving a one sentence example might be helpful here. Ie: file on the filesystem

Reviewer: Also the word specific is mispelled.

> So, for any given Temporal Workflow, how do we make sure a specific Worker is executing a specific Task? 

Reviewer: The double comma makes this sentence complex and unfriendly IMO.

Reviewer: Maybe something like, "This means the question is..."

Author: I preferred the Workflow context that the prepositional phrase provides... but OK sure.

> ## Use Task Queues

Reviewer: Why "Use"? I say drop it.

Author: The heading answers the question above it. But I see your point in terms of heading "consistency", i.e. using statements, questions, or topics as headers. I disagree that we should be consistent on that point though as it is all about accurately labeling information sections, and I want to be able to use any of them to best inform a user. Limiting ourselves like that is ... silly IMO, but will concede on this one.

> "Task Routing" is the development technique that makes use of Task Queues to orchestrate the execution of Tasks by specific Workers.

Reviewer: This sentence is very wordy and sort of cyclical.

Reviewer: Specifically " is the development technique that makes use of Task Queues" but the whole thing could probably be simpler.

Let's look at the ["Hello World" sample in Go](https://github.com/temporalio/temporal-go-samples/tree/master/helloworld) to see how this is done.

> Here, we are creating a new Worker that is subscribed to the "hello-world" Task Queue, and is registered to handle the `helloworld.Workflow` and `helloworld.Activity`:

Reviewer: I'm not a fan of "here" or "this". Also I think linking to the exact line of the sample as a hyperlink would be nice.

```go

> // The Worker is created and subscribed to the "hello-world" Task Queue

Reviewer: Don't start code comments with "The". You'll notice that almost all code comments could be started with The.

w := worker.New(c, "hello-world", worker.Options{})
defer w.Stop()
// The Worker is registered to handle the helloworld Workflow
w.RegisterWorkflow(helloworld.Workflow)
// The Worker is registered to handle the helloworld Activity
w.RegisterActivity(helloworld.Activity)
```

> And here, we are starting the `helloworld.Workflow`:

Reviewer: Again no here. Maybe, "In the next part of the code we start..."

```go
// The Workflow Options specify the "hello-world" Task Queue

> // This is the Task Queue that the Workflow will add its Tasks to

Reviewer: I think "send" instead of "add" is more accurte.

workflowOptions := client.StartWorkflowOptions{
  ID:        "hello_world_workflowID",
  TaskQueue: "hello-world",
}
// The Workflow Options are passed to the execution of the Workflow
we, err := c.ExecuteWorkflow(
  context.Background(), 
  workflowOptions, 
  helloworld.Workflow, 
  "Temporal",
)
```

> Task routing can be as simple as making sure a Worker is subscribed to a Task Queue and Tasks are being added to it.

Reviewer: I don't find that this sentence adds value.

:::caution

If a Worker pulls a Task that it has not been registered to run, an "unknown workflow type" error will occur.

:::

## Task routing for Activity dependencies

> The Temporal service will persist Activity results as long as the results can be serialized and stored in a Cassandra or MySQL database (which happens to be true for most Activity results). However, there are some scenarios where an Activity execution will do something that results in data being stored on the host where the Worker is running. It is also possible that the next Activity in the Workflow has a dependency on that data.

Reviewer: "The Temporal service will persist Activity results as long as the results can be serialized and stored in a Cassandra or MySQL database (which happens to be true for most Activity results)"

Few things:
* Dont think Cassandra or MySQL are relevant here. They are really a technical detail. The limits are 2mb and must be serializable, datastore doesn't matter
* "Persist Activity results" isn't wrong but it's sort of not the point you're trying to make. It might be worth rewording to use the term "event history".

```suggestion
The Temporal service will persist Activity results as long as the results can be serialized and stored in a Cassandra or MySQL database (which happens to be true for most Activity results). However, in some cases you may have an Activity which creates large files which are stored directly on the host file system of its worker. If the next activity in the workflow depends on that same file it will need to be run on the same host.
```

Author: AFAIK, Events are not the same as Activity results. So I don't think that would make sense.

> Let's look at a file processing Workflow, as an example,  where files are downloaded, processed in some way, and then uploaded to some other location. If a large video file is being processed only the name of the file can persist normally. The file itself must persist locally on the Worker host. This means that any future Activities that act on the video file must also take place on the host where the Worker downloaded it.

Reviewer: ```suggestion
Let's look at a concrete file processing Workflow example where  where files are downloaded, processed in some way, and then uploaded to some other location. If a large video file is being processed only the name of the file can persist normally. The file itself must persist locally on the Worker host. This means that any future Activities that act on the video file must also take place on the host where the Worker downloaded it.
```

I think you should focus more on what limits the activities from passing the file back. You just sort of skip over that part.

Author: I am hesitant to use "concrete" as an adjective anywhere in this documentation.
Concrete: adjective
existing in a material or physical form; not abstract.

While the code sample is indeed a comprehensive organization of these components at work - it is far from "concrete" in its existence.

> So, how do we ensure that a Worker on the correct host will execute the next set of Activities that depends on the video file?

Reviewer: "set of Activities" makes it confusing, just talk about a single next Activity.

> Each SDK offers a slightly different way to handle this.

Reviewer: "Each SDK" is sort of a negative reflection on our product. I would say, "This is accomplished in different ways depending on the SDK".

### Java

> In Java the paradigm is to create two Workers. Each will exist on the same host and thus have access to the same file system, but one will listen to a shared Task Queue, which it will poll to start the Workflow, and the other will listen to its own private Task Queue.

Reviewer: "paradigm" meh - best practice maybe

Reviewer: "exist" isn't the right word.

Also the "poll to start the Workflow" part isn't super easy to grasp

Author: > "paradigm" meh - best practice maybe

paradigm (noun): a typical example or pattern of something; a model.
best practice (noun): commercial or professional procedures that are accepted or prescribed as being correct or most effective.

It is much more of a pattern/model - there is really no other alternative to make this happen. Paradigm is by far a better word choice here IMO. Best practice suggests there a bunch of ways to do it - but through experience it is considered the better way... which just is not the case here. This is the pattern that Temporal forces a developer to adhere to -

```java
// Initialize a WorkerFactory
WorkerFactory factory = WorkerFactory.newInstance(client);

> // Create a Worker that is subscribed to a shared TaskQueue

Reviewer: `TaskQueue` one word or two?

final Worker workerForCommonTaskList = factory.newWorker(globalTaskQueue);
// Register Workflow and Activities
// Create a Worker that is subscribed to a private TaskQueue
final Worker workerForHostSpecificTaskList = factory.newWorker(hostSpecifiTaskQueue);
// Register activities
```

> Then, design your Workflow Activities to return an object that includes both the Activity result and the name of the private Task Queue to which the next Activity Task will be added.

Reviewer: I wouldn't say "then design your"

I think it might be worth mentioning that the first activity will run on arbitrary host and pass back the host for the subsequent activities

```java
public interface StoreActivities {

  final class ResultAndTaskQueuePair {
    // This is the name of the private Task Qeueu
    private final String hostspecificTaskQueue;
    // This is the Activity result (e.g. the name or location of the file)
    private final String activityResult;

    // Get and set methods ...
  }

  // Interface methods ....
}
```

> The [FileProcessing sample in Java](https://github.com/temporalio/temporal-java-samples/tree/143902cba3e14aa11f3b90784baf53406c329fd2/src/main/java/io/temporal/samples/fileprocessing) shows an example of this design.

Reviewer: This link could be dangerous. Why isn't it tied to master?

### Go

> In Go, you actually have two options.

Reviewer: ```suggestion
In Go, you have two options.
```

> The first option is to do the same thing as you would do in Java, which is to create two Workers (one listens to the shared Task Qeue and the other listens to the private Task Queue) and return the name of the private Task Queue back to the Workflow as an Activity result.

Reviewer: ```suggestion
**Option 1**: same design as Java, which is to create two Workers (one listens to the shared Task Qeue and the other listens to the private Task Queue) and return the name of the private Task Queue back to the Workflow as an Activity result.
```

> The second option is to create a Workflow Session which automatically ties any remaining Workflow Activity executions to the Worker that started the Workflow.

Reviewer: ```suggestion
**Option 2**: create a Workflow Session which automatically ties any remaining Workflow Activity executions to the Worker that started the Workflow.
```

---
id: route-tasks
title: How to route Tasks
sidebar_label: Route Tasks
---

Consider the specifications of a Worker:

> - A Worker only handles [Activity](docs/learn-activities) and [Workflow](docs/learn-workflows) Tasks that it is registered to handle. This means that a Task must be paired with a Worker that is registered to handle that Task.

Author: Yes, but that is done in the very next section.

- A Worker runs on a specific host. This means that Activity Tasks that depend on host-specific resources, such as a local file, must be executed by a Worker on that host.

The question is how do we make sure a specific Worker is executing a specific Task? We do this through "Task routing".

## Task routing with Task Queues

> Task Queues are what enable "Task routing". Task routing is the development technique that pairs Tasks with Workers. The basics of it can be broken down into 3 steps:

Reviewer: ```suggestion
Task routing is the development technique that pairs Tasks with Workers using Task Queues. At a basic level it consists of 3 steps:
```

> 1. Create a Worker that is registered to handle Workflow and/or Activity Tasks.

Reviewer: I think this should either be reworded or split into two steps.

1. Create worker
2. Register worker for handle...

You could also just keep Register and drop Create as its pretty heavily implied

> 2. Subscribe the Worker to listen to a Task Queue.

Reviewer: "Subscribe to listen" is sort of a weird phrasing

> 3. Configure Workflows and Activities to send Tasks to the Task Queue.

Reviewer: It might be nice to either reference `ActivityOptions`/`WorkflowOptions` or say "configure activity execution"

Author: My goal with this was to explain what is happening, and then in the following example show how it happens. Will think about this -

> Let's look at the ["Hello World" sample in Go](https://github.com/temporalio/temporal-go-samples/tree/master/helloworld) to see how this is done.

Reviewer: I would combine this into a single paragraph with the next section

> In the following code snippet, we [create a new Worker](https://github.com/temporalio/go-samples/blob/master/helloworld/worker/main.go#L26) that is subscribed to the "hello-world" Task Queue and is registered to handle the `helloworld.Workflow` and `helloworld.Activity`:

Reviewer: It might be nice to say where this code resides. Ie: Starter, worker, workflow, activity etc. Also

"create a new worker and subscribe it", sayings "that is subscribed" implies this is done automatically as opposed to in the below code.

Reviewer: also "handle the ...  `Activity` implementations"

```go
// Worker is created and subscribed to the "hello-world" Task Queue
w := worker.New(c, "hello-world", worker.Options{})
defer w.Stop()
// Worker is registered to handle the helloworld Workflow
w.RegisterWorkflow(helloworld.Workflow)
// Worker is registered to handle the helloworld Activity
w.RegisterActivity(helloworld.Activity)
```

> In the next code snippet, we [start the `helloworld.Workflow`](https://github.com/temporalio/go-samples/blob/master/helloworld/starter/main.go#L25):

Reviewer: Again where would this reside

Author: Where it resides definitely depends on the SDK language and app design - I would like to experiment with using a "Sample source reference" link that would send the user to the exact code lines in the sample that best reflects the snippet (the actual go samples need to be updated to be better samples for the docs - part of the sync project)

```go
// Workflow Options specify the "hello-world" Task Queue
// Which is the Task Queue that the Workflow will send its Tasks to
workflowOptions := client.StartWorkflowOptions{
  ID:        "hello_world_workflowID",
  TaskQueue: "hello-world",
}
// Workflow Options are passed to the execution of the Workflow
we, err := c.ExecuteWorkflow(
  context.Background(), 
  workflowOptions, 
  helloworld.Workflow, 
  "Temporal",
)
```

:::caution

If a Worker pulls a Task that it has not been registered to run, an "unknown workflow type" error will occur.

:::

## Task routing for Activity dependencies

> The Temporal service will persist Activity results as long as the results are less than 2MB in size and can be serialized for storage (which happens to be true for most Activity results). However, there are some scenarios where an Activity execution will do something that results in data, such as a large file, being stored directly on the host where the Worker is running. It is also possible that the next Activity in the Workflow has a dependency on that data and needs to run on the same host.

Reviewer: I just wanted to say that this is an exceptional first sentence.

Reviewer: "However, there are some scenarios where an Activity execution will do something that results in data, such as a large file, being stored directly on the host where the Worker is running."

results in data, such as a large file, being directly stored

I don't think data is a large file which makes this weird, the comma after "large file" makes it even weirder.

> Let's look at a file processing Workflow example where files are downloaded, processed in some way, and then uploaded to some other location. If a large video file is being processed only the name of the file can persist normally. The file itself must persist locally on the Worker host, as it is too big for normal Temporal storage. This means that any future Activities that act on the video file must also take place on the host where the Worker downloaded it.

Reviewer: I think it might make sense to frontload this before the actual code etc. Right now this doc is very abstract until this section. I understand its a bit tricky because the problem itself isn't clear until you understand the limit on activities etc.

Reviewer: Also "normal Temporal storage" is too generic

Reviewer: "processed" with "transcoded"

> So, how do we ensure that a Worker on the correct host will execute the next Activity that depends on the video file?

Reviewer: Combine these two sentences into one paragraph

This is accomplished in different ways depending on the SDK.

### Java

> Consider the [file processing sample in Java](https://github.com/temporalio/java-samples/tree/master/src/main/java/io/temporal/samples/fileprocessing). The paradigm is to create two Workers. Each will run on the same host and thus have access to the same file system, but one will listen to a global (shared) Task Queue, which it will use to start the Workflow, and the other will listen to its own host-specific (private) Task Queue. 

Reviewer: This is a very good explanation. It might be worth briefly commenting on how "host-specific (private)" is accomplished (unique host id/mac address etc)

The following code snippet is stripped down from the [FileProcessingWorker.java sample](https://github.com/temporalio/java-samples/blob/master/src/main/java/io/temporal/samples/fileprocessing/FileProcessingWorker.java) and highlights the creation of two workers for this purpose:

```java
// Initialize a WorkerFactory
WorkerFactory factory = WorkerFactory.newInstance(client);
// Create a Worker that is subscribed to a shared Task Queue
final Worker workerForCommonTaskList = factory.newWorker(globalTaskQueue);
// Register Workflow and Activities
// Create a Worker that is subscribed to a private TaskQueue
final Worker workerForHostSpecificTaskList = factory.newWorker(hostSpecifiTaskQueue);
// Register activities
```

> In the [Workflow](https://github.com/temporalio/java-samples/blob/master/src/main/java/io/temporal/samples/fileprocessing/FileProcessingWorkflowImpl.java#L42), the first Activity Task is executed by the Worker listening to the global Task Queue:

Reviewer: I think you could make it clear that the task is executed by that worker because the activity was executed using that tasklist.

```java
public FileProcessingWorkflowImpl() {
    // Create activity clients.
    ActivityOptions ao =
        ActivityOptions.newBuilder()
        // Send the Activity to the shared Task Queue
        .setTaskQueue(globalTaskQueue)
        .build();
    this.defaultTaskQueueStore = Workflow.newActivityStub(StoreActivities.class, ao);
}
```

> The Activities should be designed to [return an object](https://github.com/temporalio/java-samples/blob/master/src/main/java/io/temporal/samples/fileprocessing/StoreActivities.java) that includes both the Activity result and the name of the host-specific Task Queue to which the next Activity Task will be sent.

Reviewer: ```suggestion
The Activities should be designed to [return an object](https://github.com/temporalio/java-samples/blob/master/src/main/java/io/temporal/samples/fileprocessing/StoreActivities.java) that includes both the Activity result and the name of the host-specific Task Queue to which the next Activity Task should be sent.
```

```java
public interface StoreActivities {

  final class ResultAndTaskQueuePair {

>     // Name of the host-specific  Task Qeueu

Reviewer: Qeueu

    private final String hostspecificTaskQueue;
    // This is the Activity result (e.g. the name or location of the file)
    private final String activityResult;

    // Get and set methods ...
  }

  // Interface methods ....
}
```

> The host-specific Task Queue is then set on [subsequent Activities](https://github.com/temporalio/java-samples/blob/master/src/main/java/io/temporal/samples/fileprocessing/FileProcessingWorkflowImpl.java#L63):

Reviewer: '"set" on subsequent activites'

I think this isn't the most accurate way of phrasing it. Activities aren't really set, the are configured when executed.

```java
private void processFileImpl(URL source, URL destination) {
    // Initialize stubs that are specific to the returned task queue.
    ActivityOptions hostActivityOptions =
        ActivityOptions.newBuilder()
        // Set the host-specific Task Queue
        .setTaskQueue(hostSpecificTaskQueue)
        .build();

    // Call the Activity to process the file using worker-specific task queue.
}
```

### Go

In Go, you have two options.

> **Option 1**: Implement the same design as you would in Java, which is to create two Workers (one listens to the shared Task Qeue and the other listens to the private Task Queue) and return the name of the private Task Queue back to the Workflow as an Activity result.

Reviewer: Very concise and easy to read sentence.

> **Option 2**: Create a Workflow Session which automatically ties any remaining Workflow Activity executions to the Worker that started the Workflow. The [FileProcessing sample in Go](https://github.com/temporalio/temporal-go-samples/tree/master/fileprocessing) is a good example of sessions being used for this purpose.

Reviewer: "any remaining Workflow Activity executions "

I think "remaining" is misleading

> The Worker must be [configured to support sessions](https://github.com/temporalio/go-samples/blob/master/fileprocessing/worker/main.go#L27).

Reviewer: This should be part of the previous para

```go
workerOptions := worker.Options{
  // Sessions are enabled in the WorkerOptions
  EnableSessionWorker: true,
}
// The Worker is subscribed to a Task Queue
w := worker.New(c, "task_queue_name", workerOptions)
defer w.Stop()
```

Within the [Workflow invocation code](https://github.com/temporalio/go-samples/blob/master/fileprocessing/workflow.go#L48), set the `SessionOptions` and create the Session.

```go
// The Workflow SessionOptions are configured
so := &workflow.SessionOptions{
  CreationTimeout:  time.Minute,
  ExecutionTimeout: time.Minute,
}
// The Workflow Session is created
sessionCtx, err := workflow.CreateSession(ctx, so)
if err != nil {
  return err
}
defer workflow.CompleteSession(sessionCtx)
// Execute activities...
```

The session will ensure that all Activities are executed on the same host.

> 

Reviewer: It might be nice to write a brief conclusion

---
id: route-tasks
title: How to route Tasks
sidebar_label: Route Tasks
---

## Overview

This guide covers the recommended techniques to ensure that a Workflow and/or Activity will be executed by an appropriate Worker. In this guide we will look at two examples:

1. A basic example that illustrates the fundamentals of routing a Task to a Worker.
2. A more complex example involving Activities with host-specific file dependencies.

Consider the specifications of a Worker:
- A Worker only handles [Activity](activities) and [Workflow](workflows) Tasks that it is registered to handle. This means that a Task must be paired with a Worker that is registered to handle that Task.

> - A Worker runs on a specific host. This means that Activity Tasks that depend on host-specific resources, such as a local file, must be executed by a Worker on that host.

Reviewer 2: Actually a worker runs in a specific process. And it is possible for an activity to depend on process-specific resources like in-memory cache.

    ]           
}>

<TabItem value="java">

[Sample source reference](https://github.com/temporalio/java-samples/blob/master/src/main/java/io/temporal/samples/hello/HelloActivity.java#L91-L99)
```java
HelloWorldWorkflow workflow =
    client.newWorkflowStub(
        HelloWorldWorkflow.class,
        // Workflow Options specify the "hello-world"" Task Queue
        // Which is the Task Queue that the Workflow will send its Tasks to
        WorkflowOptions.newBuilder().setTaskQueue("hello_world").build());
// Workflow is executed
WorkflowExecution workflowExecution =
    WorkflowClient.start(workflow::helloWorld);
```

</TabItem>
<TabItem value="go">

[Sample source reference](https://github.com/temporalio/go-samples/blob/master/helloworld/starter/main.go#L20-L25)
```go
// Workflow Options specify the "hello-world" Task Queue
// Which is the Task Queue that the Workflow will send its Tasks to
workflowOptions := client.StartWorkflowOptions{
  ID:        "hello_world_workflowID",
  TaskQueue: "hello-world",
}
// Workflow Options are passed to the execution of the Workflow
we, err := c.ExecuteWorkflow(
  context.Background(), 
  workflowOptions, 
  helloworld.Workflow, 

>   "Temporal",

Reviewer 2: This "Temporal" parameter is confusing, especially having that the Java sample one doesn't have it.

)
```

</TabItem>
</Tabs>

:::caution

> If a Worker pulls a Task that it has not been registered to execute, an "unknown workflow type" error will occur.

Reviewer 2: We should add that it implies that all workers that listen to a given task queue should have the same set of workflows and activities registered with the corresponding workers.

:::

The above examples illustrate the fundamental approach to pairing Tasks with Workers. In the next section we look at a more complex scenario involving Activities with host-specific file dependencies.

> 

Reviewer 2: Host specific task routing is a very important use case. But the more important use case is running different activities and workflows in different services (which are essentially pools of processes). I think we want to have a section on using task queues to route to different service/deployment.

```java
private void processFileImpl(URL source, URL destination) {
    // Initialize stubs that are specific to the returned task queue.
    ActivityOptions hostActivityOptions =
        ActivityOptions.newBuilder()
        // Set the host-specific Task Queue
        .setTaskQueue(hostSpecificTaskQueue)
        .build();

    // Call the Activity to process the file using worker-specific task queue.
}
```

### Go

In Go, you have two options.

**Option 1**: Implement the same design as you would in Java, which is to create two Workers (one listens to the shared Task Qeue and the other listens to the private Task Queue) and return the name of the private Task Queue back to the Workflow as an Activity result.

**Option 2**: Create a Workflow Session which automatically ties Activity executions to the Worker that started the Workflow. The [FileProcessing sample in Go](https://github.com/temporalio/temporal-go-samples/tree/master/fileprocessing) is a good example of sessions being used for this purpose. The Worker must be [configured to support sessions](https://github.com/temporalio/go-samples/blob/master/fileprocessing/worker/main.go#L27).

> 

Reviewer 2: I think it is worth adding that session allows limiting number of parallel sessions per process.
